crates/subspace-farmer/src/dsn.rs

```diff
@@ -50,101 +88,75 @@ where
     DSN: DSNSync + Send + Sized,
     DSN::Stream: Unpin + Send,
     // On pieces might `break` with some result from the sync or just continue
-    OP: FnMut(FlatPieces, Vec<PieceIndex>) -> ControlFlow<anyhow::Result<()>> + Send + 'static,
+    OP: FnMut(FlatPieces, Vec<PieceIndex>) -> anyhow::Result<()> + Send + 'static,
 {
     let SyncOptions {
+        max_plot_size,
+        total_pieces,
         range_size,
         address,
     } = options;
-    let mut increasing_cursor = PieceIndexHashNumber::from(Sha256Hash::from(address));
-    let mut decreasing_cursor = increasing_cursor;
+    let address = PieceIndexHashNumber::from(Sha256Hash::from(address));
+    let max_distance_one_direction = if total_pieces < max_plot_size {
+        PieceIndexHashNumber::MAX / 2
+    } else {
+        PieceIndexHashNumber::MAX / total_pieces * max_plot_size / 2
+    };
+    let (from, is_underflow) = address.overflowing_sub(max_distance_one_direction);
+    let (end, is_overflow) = address.overflowing_add(max_distance_one_direction);
     let on_pieces = Arc::new(Mutex::new(on_pieces));
-    let stop_at = increasing_cursor.wrapping_add(&(PieceIndexHashNumber::MAX / 2));
 
-    while increasing_cursor < stop_at || decreasing_cursor > stop_at {
-        let increasing_stream: Box<dyn Stream<Item = PiecesToPlot> + Unpin + Send> = {
-            let start = increasing_cursor;
+    // Sync first half of plot
```

Comment:
Once you have identified `from` and `to`, create a list of non-overflowing ranges (there will be either one or two of these). Then you can iterate over them and sync smaller non-overflowing sectors.

Currently you have several places that potentially overflow and keeping all of those combinations in mind is harder than it needs to be.

Also once you have a loop over non-overlapping ranges and a loop inside of that to pull pieces, you don't need `sync_sector` and can inline it here, which saves you some generic code and improves readability further.

Reply:
Hm, if we change API from `from` + `to` to `from`+`count` we don't need to worry about overflow at all, we can just move from `from` by requesting up to `N` pieces in every request until we get to or cross `to`. I think it is a nice approach because none of the overflows would have to be handled, all we need to do is to use `wrapping_add`.